### PR TITLE
Solidity to Yul compiler

### PR DESCRIPTION
**Pull request by chriseth**
Partly replaces https://github.com/ethereum/solidity/pull/2320

Depends on https://github.com/ethereum/solidity/pull/6396 and #6402

Implements the bare minimum to generate the IR of a contract. Does not yet have all the "unimplemented feature" guards. We might still want to merge it in this state after a thorough design discussion.

### CODE REVIEW

**Comment by chriseth** [comment]
Added full dispatch and decoding.

**Comment by chriseth** [comment]
This might cause problems (failed assertions, etc) if we select an output of `*` in standard-json, which would include yul. Shall we exclude `yul` from `*`?

**Comment by ekpyron** [comment]
Would be easier to review if it were rebased...

File: libsolidity/interface/StandardCompiler.cpp [hunk 1]
```diff
@@ -208,6 +209,23 @@ bool isBinaryRequested(Json::Value const& _outputSelection)
 	return false;
 }
 
+/// @returns true if any Yul IR was requested. Note that as an exceptin, '*' does not
+/// yet match "ir" or "irOptimized"
+bool isIRRequested(Json::Value const& _outputSelection)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp:220]: The beauty of this function is not even topped by the most sophisticated async javascript code I have seen so far ;)

**Comment by chriseth** [comment]
When using the optimizer, it compiles as follows:
```
contract C {
  function f(uint a, uint b, bool c) public pure returns (uint x, uint y, bool z) {}
}
```
->
```
object "C_18" {
    code {
        let _1 := datasize("C_18_deployed")
        codecopy(0, dataoffset("C_18_deployed"), _1)
        return(0, _1)
    }
    object "C_18_deployed" {
        code {
            if iszero(lt(calldatasize(), 4))
            {
                let _1 := 0
                if eq(0x80330ef1, shr(224, calldataload(_1)))
                {
                    if callvalue()
                    {
                        revert(_1, _1)
                    }
                    if slt(add(calldatasize(), 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc), 96)
                    {
                        revert(_1, _1)
                    }
                    let memPtr := mload(64)
                    if gt(memPtr, 0xffffffffffffffff)
                    {
                        revert(_1, _1)
                    }
                    mstore(64, memPtr)
                    mstore(memPtr, _1)
                    mstore(add(memPtr, 32), _1)
                    mstore(add(memPtr, 64), _1)
                    return(memPtr, 96)
                }
            }
            revert(0, 0)
        }
    }
}
```

I think it could use some dead store elimination, but apart from that, pretty nice!
(nb: the full ABI encoder and decoder code is generated and then optimized away)

**Comment by chriseth** [comment]
Ah, just noticed: the memory pointer init code is missing.

File: libsolidity/interface/StandardCompiler.cpp [hunk 2]
```diff
@@ -208,6 +209,23 @@ bool isBinaryRequested(Json::Value const& _outputSelection)
 	return false;
 }
 
+/// @returns true if any Yul IR was requested. Note that as an exceptin, '*' does not
```
Comment (leonardoalt) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
/// @returns true if any Yul IR was requested. Note that as an exception, '*' does not
```

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+{
+	return _contract.name() + "_" + to_string(_contract.id());
+}
+
+string IRGenerator::runtimeObjectName(ContractDefinition const& _contract)
+{
+	return _contract.name() + "_" + to_string(_contract.id()) + "_deployed";
+}
+
+string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Note to self: still need to review this function.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+		t("fallback", fallbackCode);
+	}
+	else
+		t("fallback", "revert(0, 0)");
+	return t.render();
+}
+
+string IRGenerator::memoryInit()
+{
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGenerator.cpp:224]: Maybe a comment about memory being zero-initialized and not having to write to the zero pointer wouldn't hurt.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 3 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerator.cpp]: I guess we postpone things like binary dispatch? Probably makes sense.

**Review by ekpyron — COMMENTED** [review]
Looks good so far

**Comment by chriseth** [comment]
@ekpyron weird, I cannot "answer" to your [comment](https://github.com/ethereum/solidity/pull/6179#discussion_r271223574). My idea would be to make binary search for dispatch just a generic optimizer step for all switch statements.

**Comment by ekpyron** [comment]
@chriseth Yeah, that makes sense. EDIT: ah - I think because while reviewing I commented on leo's comment, my comment is now there twice and one copy seems broken :-)...

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+				// <functionName>
+				<callValueCheck>
+				<assignToParams> <abiDecode>(4, calldatasize())
+				<assignToRetParams> <function>(<params>)
+				let memPos := <allocate>(0)
+				let memEnd := <abiEncode>(memPos <comma> <retParams>)
+				return(memPos, sub(memEnd, memPos))
+			}
+			</cases>
+			default {}
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGenerator.cpp:179]: Why is this not the same as the fallback?

**Comment by leonardoalt** [comment]
Can you add a test with parameters and return parameters?

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 5 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerator.cpp:179]: It is... this doesn't return, but will instead continue with ``<fallback>`` below, won't it?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp:179]: Yep.

**Comment by chriseth** [comment]
Updated - mainly added the warning, the zero memory comment, added an assertion in the constructor code and corrected typos.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 5 above)
Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGenerator.cpp:179]: oh, right

**Comment by chriseth** [comment]
@leonardoalt the tests are already in #6403 - do you want me to add those here?

**Comment by chriseth** [comment]
fuzzer is failing.

**Comment by chriseth** [comment]
Yeah, probably a good idea.

**Comment by leonardoalt** [comment]
Looks good. Do we want to wait for the CI bugfix PR to be merged?

**Comment by chriseth** [comment]
Fixed the problems with the commandline tests. I would merge it if we have one good run of the commandline tests and of a single configuration combination of soltest.

**Review by leonardoalt — APPROVED** [review]

**Review by ekpyron — APPROVED** [review]
Still haven't reviewed it in all detail, but I think it's fine anyways.

**Pull request merged**